modnet/models.py

```diff
@@ -326,60 +344,122 @@ def fit_preset(
             n_feat = min(len(data.get_optimal_descriptors()), params["n_feat"])
-            self.model = MODNetModel(
-                self.targets,
-                self.weights,
-                num_neurons=params["num_neurons"],
-                n_feat=n_feat,
-                act=params["act"],
-                num_classes=self.num_classes
-            ).model
-            self.n_feat = n_feat
-            self.fit(
-                data,
-                val_fraction=val_fraction,
-                lr=params["lr"],
-                epochs=params["epochs"],
-                batch_size=params["batch_size"],
-                loss=params["loss"],
-                callbacks=callbacks,
-                verbose=verbose,
-            )
-            val_loss = np.array(self.model.history.history["val_loss"])[-20:].mean()
+
+            splits = matbench_kfold_splits(data, classification=classification)
+            if not nested:
+                splits = [next(splits)]
+
+            nested_val_losses = []
+            for ind, (train, test) in enumerate(splits):
+                LOG.info("Initialising split #{} preset #{}/{}: {}".format(ind + 1, i + 1, len(presets), params))
+
+                val_params = {}
+                if nested:
+                    train_data, test_data = data.split((train, test))
+                    val_params["val_data"] = test_data
+                else:
+                    val_params["val_fraction"] = 0.15
+                    train_data = data
+
+                self.model = MODNetModel(
+                    self.targets,
+                    self.weights,
+                    num_neurons=params["num_neurons"],
+                    n_feat=n_feat,
+                    act=params["act"],
+                    num_classes=self.num_classes
+                ).model
+                self.n_feat = n_feat
+                self.fit(
+                    train_data,
+                    lr=params["lr"],
+                    epochs=params["epochs"],
+                    batch_size=params["batch_size"],
+                    loss=params["loss"],
+                    callbacks=callbacks,
+                    verbose=verbose,
+                    **val_params
+                )
+
+                learning_curve = self.model.history.history["val_loss"]
+
+                if any(isinstance(cb, keras.callbacks.EarlyStopping) for cb in callbacks):
+                    val_loss = np.min(learning_curve)
+                    mean_val_loss = np.mean(learning_curve[-20:])
+                    LOG.info(f"Taking loss {val_loss:3.3f} from EarlyStopping vs mean {mean_val_loss:3.3f}")
+                else:
+                    LOG.info(f"Loss from curve {val_loss:3.3f}")
+                    val_loss = np.mean(learning_curve[-20:])
+
+                nested_val_losses.append(val_loss)
+
+            val_loss = np.mean(nested_val_losses)
+
             if val_loss < min(val_losses):
                 best_model = self.model
```

Comment:
we should in fact have a callback picking up the model with lowest validation loss